common/tools/dev-tool/src/commands/samples/publish.ts

```diff
@@ -0,0 +1,615 @@
+          }
+        }
+        const tags = ts.getJSDocTags(node);
+
+        // Look for the @summary jsdoc tag block as well
+        if (summary === undefined && tags.some(({ tagName: { text } }) => text === "summary")) {
+          for (const tag of tags) {
+            if (tag.tagName.text === "summary") {
+              log.debug("Found summary tag on node:", node.getText(sourceFile));
+              // Replace is required due to multi-line splitting messing with table formatting
+              summary = tag.comment?.replace(/\s*\n\s*/, " ");
+            } else if (
+              tag.tagName.text.startsWith(`${AZSDK_META_TAG_PREFIX}`) &&
+              tag.comment !== undefined
+            ) {
+              // We ran into an `azsdk` directive in the metadata
+              const metaTag = tag.tagName.text.replace(
+                new RegExp(`^${AZSDK_META_TAG_PREFIX}`),
+                ""
+              ) as keyof AzSdkMetaTags;
+              if (validAzSdkTags.includes(metaTag)) {
+                azSdkTags[metaTag as keyof AzSdkMetaTags] = JSON.parse(tag.comment);
+              } else {
+                log.warn(`Invalid azsdk tag ${metaTag}. Valid tags include ${validAzSdkTags}`);
+              }
+            }
+          }
+        }
+
+        ts.visitEachChild(node, visitor, context);
+        return node;
+      };
+
+      ts.visitNode(sourceFile, visitor);
+      return sourceFile;
+    };
+
+    return {
+      filePath: source,
+      relativeSourcePath,
+      text: sourceText,
+      jsModuleText: convert(sourceText, {
+        fileName: source,
+        transformers: {
+          before: [sourceProcessor]
+        }
+      }),
+      summary: summary ?? fail(`${relativeSourcePath} does not include an @summary tag.`),
+      importedModules: importedModules.filter((name) => !nodeBuiltins.includes(name)),
+      usedEnvironmentVariables,
+      azSdkTags
+    };
+  });
+
+  return Promise.all(jobs);
+}
+
+/**
+ * Extracts the sample generation metainformation from the sample sources and
+ * configuration in package.json.
+ *
+ * This is the function that assembles all the information that the templates
+ * use to generate good output.
+ */
+async function getSampleGenerationInfo(
+  projectInfo: ProjectInfo,
+  topLevelDirectory: string,
+  onError: () => void
+): Promise<SampleGenerationInfo> {
+  const sampleSourcesPath = path.join(projectInfo.path, DEV_SAMPLES_BASE);
+  const sampleSources = (await fs.readdir(sampleSourcesPath))
+    .filter((name) => name.endsWith(".ts"))
+    .map((name) => path.join(sampleSourcesPath, name));
+
+  const sampleConfiguration = getSampleConfiguration(projectInfo.packageJson);
+
+  const baseName = projectInfo.name.split("/").slice(-1)[0];
+
+  log.debug("Determined project baseName:", baseName);
+
+  // A helper to handle configuration errors.
+  function fail(...values: unknown[]): never {
+    log.error(...values);
+    onError();
+    return undefined as never;
```

Comment:
why not just have this method have a void return? Is it because onError is expected to exit?

Reply:
This allows me to use `fail(...)` in an expression/assignment context where I'm sure the result won't ever be accessed. E.g. 

```javascript
return {
  field: conditionForSuccess ? actualValue : fail([
    "This will return undefined, but be assignable because it's `never`-typed.",
    "I know it's safe because `fail` causes an exception before anything can read this field."
  ].join("\n"))
};
```

It's not the most elegant design, but was one of the only ways I could think of to avoid throwing immediately and allowing as much of the program to proceed as possible without infecting the entire type ecosystem with optional fields that will never actually be undefined in a non-error case.

Reply:
This makes sense, it's a pretty elegant solution actually